### DISCUSSION TITLE
error flutter_rust_bridge and dart to compile librustdesk.so

### DISCUSSION THREAD

**Discussion by TheBrokenCoder** [rustdesk/rustdesk#5561] [Q&A]
Hello.

I followed the official documentation on:
https://rustdesk.com/docs/en/dev/build/android/

but when i try to compile librustdesk.so running the command : 
./flutter/ndk_arm64.sh

i have this error output:

-------------------------------------------------------------------------------------------------------------------------------------------------
Building arm64-v8a (aarch64-linux-android)
warning: unreachable statement
   --> build.rs:129:5
    |
123 |   return;
    |   ------ any code following this expression is unreachable
...
129 |   let target_os = std::env::var("CARGO_CFG_TARGET_OS").unwrap();
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unreachable statement
    |
    = note: `#[warn(unreachable_code)]` on by default

warning: function `gen_flutter_rust_bridge` is never used
  --> build.rs:80:4
   |
80 | fn gen_flutter_rust_bridge() {
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: `rustdesk` (build script) generated 2 warnings
   Compiling rustdesk v1.2.3 (/home/ubuntu/rustdesk)
error[E0432]: unresolved import `flutter_rust_bridge::rust2dart::IntoIntoDart`
  --> src/bridge_generated.rs:16:5
   |
16 | use flutter_rust_bridge::rust2dart::IntoIntoDart;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------
   |     |                               |
   |     |                               help: a similar name exists in the module: `IntoDart`
   |     no `IntoIntoDart` in `rust2dart`

error[E0277]: the trait bound `DartCObject: From<EventToUI>` is not satisfied
   --> src/flutter.rs:792:19
    |
792 |     event_stream: StreamSink<EventToUI>,
    |     ^^^^^^^^^^^^^^^^^^^^^ the trait `From<EventToUI>` is not implemented for `DartCObject`
    |
    = help: the following other types implement trait `From<T>`:
              <DartCObject as From<RustOpaque<T>>>
              <DartCObject as From<DartOpaque>>
    = note: required for `EventToUI` to implement `Into<DartCObject>`
    = note: required for `EventToUI` to implement `IntoDart`
note: required by a bound in `StreamSink`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/flutter_rust_bridge-1.75.3/src/rust2dart.rs:97:26
    |
97  | pub struct StreamSink<T: IntoDart> {
    | ^^^^^^^^ required by this bound in `StreamSink`

error[E0277]: the trait bound `DartCObject: From<EventToUI>` is not satisfied
   --> src/flutter_ffi.rs:110:16
    |
110 |     events2ui: StreamSink<EventToUI>,
    |     ^^^^^^^^^^^^^^^^^^^^^ the trait `From<EventToUI>` is not implemented for `DartCObject`
    |
    = help: the following other types implement trait `From<T>`:
              <DartCObject as From<RustOpaque<T>>>
              <DartCObject as From<DartOpaque>>
    = note: required for `EventToUI` to implement `Into<DartCObject>`
    = note: required for `EventToUI` to implement `IntoDart`
note: required by a bound in `StreamSink`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/flutter_rust_bridge-1.75.3/src/rust2dart.rs:97:26
    |
97  | pub struct StreamSink<T: IntoDart> {
    | ^^^^^^^^ required by this bound in `StreamSink`

error[E0277]: the trait bound `DartCObject: From<EventToUI>` is not satisfied
    --> src/flutter_ffi.rs:1594:61
     |
1594 | pub fn plugin_register_event_stream(_id: String, _event2ui: StreamSink<EventToUI>) {
     |                                                             ^^^^^^^^^^^^^^^^^^^^^ the trait `From<EventToUI>` is not implemented for `DartCObject`
     |
     = help: the following other types implement trait `From<T>`:
              <DartCObject as From<RustOpaque<T>>>
            <DartCObject as From<DartOpaque>>
     = note: required for `EventToUI` to implement `Into<DartCObject>`
     = note: required for `EventToUI` to implement `IntoDart`
note: required by a bound in `StreamSink`
    --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/flutter_rust_bridge-1.75.3/src/rust2dart.rs:97:26
     |
97   | pub struct StreamSink<T: IntoDart> {
     |  ^^^^^^^^ required by this bound in `StreamSink`

Some errors have detailed explanations: E0277, E0432.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `rustdesk` (lib) due to 4 previous errors
note: If the build failed due to a missing target, you can run this command:
note: 
note:     rustup target install aarch64-linux-android

-------------------------------------------------------------------------------------------------------------------------------------------------

I guess it has to do with the interaction between flutter_rust_brdige and dart but after several days of trying I still haven't figured out how to fix it.

does anyone have any suggestions?

Thank you

**Comment by SergeyMy** [comment]
Instructions become obsolete faster than they are written!
use GitHub resources :)
https://rustdesk.com/docs/en/dev/build/all/

PS
in the same place, according to the Actions logs, you will understand where you deviated from the path when building on your host

**Comment by deep-soft** [comment]
1. create you're own script to build for Android on ubuntu
2. recreate environment from env:
https://github.com/rustdesk/rustdesk/blob/e88e17a4b0c79193813ded9f22b758ab7e2fa2fb/.github/workflows/flutter-build.yml#L13

3. commands from ./.github/workflows/bridge.yml
https://github.com/rustdesk/rustdesk/blob/e88e17a4b0c79193813ded9f22b758ab7e2fa2fb/.github/workflows/bridge.yml#L8
refered by:
https://github.com/rustdesk/rustdesk/blob/e88e17a4b0c79193813ded9f22b758ab7e2fa2fb/.github/workflows/flutter-build.yml#L376

4. commands from ./.github/workflows/flutter-build.yml
`# build-rustdesk-android:`
https://github.com/rustdesk/rustdesk/blob/e88e17a4b0c79193813ded9f22b758ab7e2fa2fb/.github/workflows/flutter-build.yml#L461

**Reply by SergeyMy** [reply]
@deep-soft 
colleague, you again repeat the meaning of my post, but only trying to confuse :)
answer from the link above in the docks: don't know how? - trust github and Actions from the developer.
:):):):):):):):)

**Reply by deep-soft** [reply]
@SergeyMy , I get it, but some users don't want to understand that or they trying to build locally.
It's not good but also not bad...
Apparently the pinned announce (https://github.com/rustdesk/rustdesk/discussions/763) it's not visible enough

**Comment by rustdesk** [comment]
Report to flutter_rust_bridge author please, it does not work with current stable Flutter.

**Comment by TheBrokenCoder** [comment]
Now it's work.
thanks everyone for the suggestion.

**Reply by TheBrokenCoder** [reply]
@Jwy-jump 
In my case the problem was the versions of the installed components.
I followed the files into workflows folder

**Comment by SergeyMy** [comment]
@Jwy-jump 
You didn’t even indicate what platform you are building for. the answer is obvious - if you have any difficulties, use the Actions GitHUB functionality

**Comment by sstecho** [comment]
@TheBrokenCoder 是如何解决的，我也是在编译apk的时候出现这样的问题，是要怎么解决的？

**Comment by ViniciusAndradex** [comment]
Did anyone manage to solve it?
